### PR TITLE
Make CPU inductor work with dynamic shapes

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93077

These errors were found by looking at wav2vec2

See https://github.com/pytorch/pytorch/issues/91719

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by voznesenskym — APPROVED** [review]

File: torch/_inductor/mkldnn.py [hunk 1]
```diff
@@ -351,7 +353,9 @@ def __init__(self, linear: nn.Module, input_size: list):
 
     def _update_module_params(self, linear, input_size):
         self.__dict__ = copy.deepcopy(linear.__dict__)
-        self.batch_size = int(reduce(lambda x, y: x * y, input_size) / input_size[-1])
+        self.batch_size = sym_int(
```
Comment (ngimel) [inline torch/_inductor/mkldnn.py:356]: `FloorDiv(reduce(...))`? Or `reduce(..., input_size[:-1])`? Converting truediv to int wasn't right before.

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/mkldnn.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/mkldnn.py:356]: Making it `reduce(..., input_size[:-1])` sounds simpler.

File: torch/_inductor/overrides.py [hunk 2]
```diff
@@ -87,7 +88,8 @@ def fuse_fx(gm: torch.fx.GraphModule, example_inputs):
     gm = remove_identity(gm)
     gm = fuse_conv_bn(gm)
     # do mkldnn fusion(conv(linear)+unary(binary)
-    gm = mkldnn_fuse_fx(gm, example_inputs)
+    if not dynamo_config.dynamic_shapes:
+        gm = mkldnn_fuse_fx(gm, example_inputs)
```
Comment (jgong5) [inline torch/_inductor/overrides.py:92]: What's the problem of running `mkldnn_fuse_fx` with dynamic shapes? Fusions should work with both static and dynamic shapes. We can follow up with the fix to enable the dynamic shape. cc @XiaobingSuper

Comment (ezyang) [reply torch/_inductor/overrides.py:92]: The problem is that some of the mkl ops we fuse into don't have meta formulas, like the packing stuff. Whenever you work out the formulas here we can reenable this! However when I looked at it briefly it was not obvious how to fix

Comment (jgong5) [reply torch/_inductor/overrides.py:92]: Thanks for the info, will do a follow-up fix to enable dynamic shapes here.

**Pull request closed without merging**